apollo-compiler/src/main/kotlin/com/apollostack/compiler/OperationTypeSpecBuilder.kt

```diff
@@ -1,26 +1,43 @@
 package com.apollostack.compiler
 
+import com.apollostack.compiler.ir.CodeGenerator
 import com.apollostack.compiler.ir.Field
-import com.cesarferreira.pluralize.singularize
+import com.apollostack.compiler.ir.Fragment
 import com.squareup.javapoet.TypeSpec
 import javax.lang.model.element.Modifier
 
 class OperationTypeSpecBuilder(
     val operationName: String,
-    val fields: List<Field>) {
+    val fields: List<Field>,
+    val allFragments: List<Fragment>) : CodeGenerator {
 
-  fun build(): TypeSpec {
-    return buildTypeSpecs(operationName, fields)
-  }
+  override fun toTypeSpec(): TypeSpec = buildTypeSpecs(operationName, fields, allFragments)
 
-  private fun buildTypeSpecs(currentTypeName: String, fields: List<Field>): TypeSpec {
-    val innerTypes = fields
-        .filter(Field::isNonScalar)
-        .map { buildTypeSpecs(it.responseName.capitalize().singularize(), it.fields!!) }
-    return TypeSpec.interfaceBuilder(currentTypeName)
-        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
-        .addMethods(fields.map(Field::toMethodSpec))
-        .addTypes(innerTypes)
-        .build()
-  }
+  private fun buildTypeSpecs(typeName: String, fields: List<Field>, fragments: List<Fragment>) =
+      TypeSpec.interfaceBuilder(typeName)
+          .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+          .addMethods(fields.map(Field::toMethodSpec))
+          .addTypes(innerTypes(fields))
+          .addTypes(innerFragments(fields, fragments))
+          .build()
+
+  /** Returns a list of fragment types referenced by the provided list of fields */
+  private fun innerFragments(fields: List<Field>, fragments: List<Fragment>): List<TypeSpec> =
+      fields.filter(Field::hasFragments)
+          .map { field ->
+            TypeSpec.interfaceBuilder(field.normalizedName())
+                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                .addMethod(Fragment.genericAccessorMethodSpec())
+                .addType(Fragment.genericInterfaceSpec(fragments))
```

Comment:
Yep, like I said, this is pretty basic and there's a bunch of things to improve. This is one of them 😄